### ISSUE TITLE
Reported string properties are always empty with non-root `hid` module on Linux

### ISSUE THREAD

**Issue by ntninja** [trezor/cython-hidapi#110]
On my Linux x86_64 machine, all the device string properties reported by `hid.enumerate()` (that is:  `serial_number`, `manufacturer_string` & `product_string`) are empty unless my program runs as root, even if the permissions on the corresponding `/dev/hidraw*` device are adjusted as documented. However, when using the `hidraw.enumerate()` instead, all these properties (and several more) are reported correctly even without changing permissions on the device node. Based on my understanding it shouldn't matter which API one uses in this case.

Background: For licensing reasons (shared VID/PID), I am required to check the value of the `manufacturer_string` field before attempting any use of the device, so unlike the usual cases of just matching the IDs having this data exposed is kinda important for my usage.

Software versions (Debian):

  * Linux 5.10.0-3-amd64
  * Python 3.9.1-1
  * libhidapi-hidraw 0.10.1+dfsg-1
  * libhidapi-libusb 0.10.1+dfsg-1
  * libudev 247.3-1
  * libusb 1.0.24-2
  * usbguard 1.0.0+ds-2

**Comment by ntninja** [comment]
I dumb, the UDev also(/only?) adjusts permissions on `/dev/bus/usb/XXX/YYY` and apparently `libusb` (unlike `lsusb`) isn't able to enumerate the string properties (and only those) without access to that device node!?

TL;DR: If you stumble on this, make sure *actually* have set up and applied the UDev rule, fumbling around with `/dev/hidraw*` is not enough.

**Comment by ntninja** [comment]
I do think this should be documented though, since `lsusb` displays the values just fine without any special privileges.

**Comment by prusnak** [comment]
> I do think this should be documented though, since `lsusb` displays the values just fine without any special privileges.

Well, it's not able to properly get the strings from the device. It just uses the static table that is bundled together with the software - http://www.linux-usb.org/usb.ids

**Comment by prusnak** [comment]
@ntninja please open a pull request where you document your suggestion - I'll review and merge it, thanks!

**Comment by ntninja** [comment]
Hmm, OK. But before I do that, can you tell me if there is any particular reason why the libusb based backend of libhidapi is exposed as the “primary” backend on Linux? Having spent some time with this library yesterday after I figure the above out, it appears to me that the hidraw backend is superior to the libusb one in every conceivable way:

  1. Device enumeration works reliably, even if the process has no permission to interact with the device later on (allowing for much better error messages), while providing important details like the usage page that simply aren't available when using libusb
  2. Both USB and Bluetooth HID devices are exposed (and also probably other things like HID over SPI or I²C), rather than USB only
  3. The hidraw backend uses sysfs for device enumeration and only opens the target device node on `hid.device.open{,_path}`, while libusb bypasses the hidraw API and instead communicates directly with the device using the corresponding USB device node – thereby duplicating the work the kernel already did on every `enumerate` call

**Comment by prusnak** [comment]
> But before I do that, can you tell me if there is any particular reason why the libusb based backend of libhidapi is exposed as the “primary” backend on Linux?

Honestly, I don't have a good answer to that. I started to maintain cython-hidapi a few years ago when Trezor was still using HID. I didn't do any significant changes to the library since it was forked, just occasional bug fixes. Now that Trezor doesn't use HID anymore, but WebUSB (via libusb), I have even less strong opinion on how to do things in cython-hidapi.

@matejcik Am I correct on that one what we don't use hidapi anymore to talk to Trezor devices, right? (with the exception of talking to really really old bootloaders of T1)

@benma since you are still using cython-hidapi to communicate with your device - can you please chime in and tell us whether you are OK with changing the default backend to hidraw?

**Comment by matejcik** [comment]
> @matejcik Am I correct on that one what we don't use hidapi anymore to talk to Trezor devices, right? (with the exception of talking to really really old bootloaders of T1)

This is correct, yes. The only usecase for hid is upgrading such devices to a webusb-enabled firmware.

**Comment by benma** [comment]
> @benma since you are still using cython-hidapi to communicate with your device - can you please chime in and tell us whether you are OK with changing the default backend to hidraw?

I think so. What would be the steps to test? Can the backend be set dynamically?

**Comment by prusnak** [comment]
> Can the backend be set dynamically?

Yes. Currently you can do `import hid` to import libusb version and `import hidraw as hid` to import the hidraw version.

In the future we might do `import hid` or `import hidraw as hid` to import the hidraw version (changing the default import to hidraw). And `import hidlibusb as hid` (or similar) to import the libusb version?

**Comment by mcuee** [comment]
That will be good.
